chains/chains.go

```diff
@@ -121,6 +121,53 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
 	return outputValue, nil
 }
 
+const _defaultApplyMaxNumberWorkers = 5
+
+// Apply executes the chain for each of the inputs asynchronously.
+func Apply(ctx context.Context, c Chain, inputValues []map[string]any, maxWorkers int, options ...ChainCallOption) ([]map[string]any, error) { // nolint:lll
+	if maxWorkers <= 0 {
+		maxWorkers = _defaultApplyMaxNumberWorkers
+	}
+
+	inputJobs := make(chan map[string]any, len(inputValues))
+	resultsChan := make(chan struct {
+		result map[string]any
+		err    error
+	}, len(inputValues))
+	defer close(inputJobs)
+	defer close(resultsChan)
+
+	for w := 0; w < maxWorkers; w++ {
+		go func() {
+			for input := range inputJobs {
+				res, err := Call(ctx, c, input, options...)
+				resultsChan <- struct {
+					result map[string]any
+					err    error
+				}{
+					result: res,
+					err:    err,
+				}
+			}
+		}()
+	}
+
+	for _, input := range inputValues {
+		inputJobs <- input
+	}
+
+	results := make([]map[string]any, len(inputValues))
+	for i := range inputValues {
+		r := <-resultsChan
+		if r.err != nil {
+			return nil, r.err
+		}
+		results[i] = r.result
+	}
```

Comment:
How do we know what output corresponds to a particular input? This seems to be storing results in random order. I think we should be passing the index of the input when we write to `inputJobs`.

In `mapResultsToLLMChainInputs` I think you could possibly be mapping wrong results to inputs.

I'm also assuming this `Apply` function can be used for things other than MapReduce, so we'd need to know the output for a particular input.

I could be missing something though - thoughts?

Reply:
Yes you are correct. I will make an update for this.